File: src/commands/uninstall.ts
import { spawnSync } from "node:child_process";
import { existsSync, rmSync } from "node:fs";
import pc from "picocolors";
import {
  type AgentId,
  ALL_AGENTS,
  ALL_TOOLS,
  agentInfo,
  detectAgent,
  type RunOpts,
  type ToolId,
  toolInfo,
  unwireTool,
} from "../registry.js";
import * as colors from "../util/colors.js";
import { removeWire } from "../util/manifest.js";
import * as paths from "../util/paths.js";
import { cacheDir } from "../util/paths.js";
import { Progress } from "../util/progress.js";
import { confirm, isInteractive, multiSelect, type SelectOption } from "../util/prompt.js";

/** Run the uninstall command: unwire tools from agents. */
export async function run(
  agentsFilter: AgentId[],
  toolsFilter: ToolId[],
  opts: RunOpts,
): Promise<number> {
  colors.banner("toksave", "uninstall");

  // ── Detect installed agents ──────────────────────────────
  const detected = ALL_AGENTS.filter((a) => detectAgent(a.id).installed).map((a) => a.id);

  if (detected.length === 0) {
    colors.raw("  Nothing wired.");
    console.log();
    return 0;
  }

  // ── Pick agents ─────────────────────────────────────────
  let agentIds: AgentId[];

  if (agentsFilter.length > 0) {
    agentIds = agentsFilter.filter((id) => detected.includes(id));
  } else if (opts.yes || !isInteractive()) {
    agentIds = [...detected];
  } else {
    const options: SelectOption[] = detected.map((id) => ({
      value: id,
      label: agentInfo(id).label,
      disabled: false,
      selected: true,
    }));
    const picked = await multiSelect("Select agents to uninstall toksave from", options);
    agentIds = picked as AgentId[];
  }

  if (agentIds.length === 0) {
    colors.raw("  Nothing selected.");
    console.log();
    return 0;
  }

  // ── Pick tools ──────────────────────────────────────────
  const tools: ToolId[] = toolsFilter.length > 0 ? toolsFilter : ALL_TOOLS.map((t) => t.id);

  // ── Unwire ──────────────────────────────────────────────
  const s = new Progress();
  s.start("Removing TokSave tools");
  for (const agentId of agentIds) {
    const info = agentInfo(agentId);
    s.update(`Unwiring ${info.label}`);
    for (const toolId of tools) {
      if (!opts.dryRun) {
        await unwireTool(agentId, toolId, opts);
        removeWire(agentId, toolId);
      }
      await new Promise((r) => setTimeout(r, 150)); // UX delay so progress bar is visible
    }
    s.stop(`${pc.green(colors.CHECK)} ${info.label}`);
  }

  // ── Cleanup cache + purge binaries if full removal ──────
  if (!opts.dryRun && tools.length === ALL_TOOLS.length && agentIds.length === detected.length) {
    const cache = cacheDir();
    if (existsSync(cache)) {
      try {
        rmSync(cache, { recursive: true, force: true });
      } catch {}
    }
    await purgeBinariesIfConfirmed(opts);
  }

  // ── Summary ─────────────────────────────────────────────
  console.log();
  const agentLabels = agentIds.map((id) => agentInfo(id).label);
  const toolLabels = tools.map((id) => toolInfo(id).label);
  colors.ok(`Uninstalled ${toolLabels.join(", ")} from ${agentLabels.join(", ")}.`);
  console.log();
  return 0;
}

async function purgeBinariesIfConfirmed(opts: RunOpts): Promise<void> {
  if (opts.dryRun || process.env.TOKSAVE_TEST === "1") return;
  let doPurge = opts.yes;
  if (!doPurge && isInteractive()) {
    doPurge = await confirm(
      "Also remove binaries/packages toksave installed (rtk, npm globals)?",
      false,
    );
  }
  if (!doPurge) return;

  // rtk
  try {
    const localBin = paths.localBin();
    const rtkPath = process.platform === "win32" ? `${localBin}/rtk.exe` : `${localBin}/rtk`;
    if (existsSync(rtkPath)) {
      spawnSync(rtkPath, ["init", "--uninstall"], { timeout: 5000 });
      try {
        rmSync(rtkPath, { force: true });
      } catch {}
    }
  } catch {}

  // npm globals
  for (const pkg of ["context-mode", "@colbymchenry/codegraph", "@dietrichgebert/ponytail"]) {
    try {
      spawnSync("npm", ["uninstall", "-g", pkg], { timeout: 10000 });
    } catch {}
  }
}

File: src/commands/uninstall.rs
use crate::cli::ParsedCli;
use crate::registry::{
    ALL_AGENTS, ALL_TOOLS, AgentId, ToolId, agent_info, detect_agent, tool_info, unwire_tool,
    verify_tool,
};
use crate::util::colors;
use crate::util::manifest::remove_wire;
use std::process::Command;

pub async fn run_uninstall(parsed: &ParsedCli) -> i32 {
    colors::banner("toksave", "uninstall");

    // ── Detect installed agents ──
    let detected: Vec<AgentId> = ALL_AGENTS
        .iter()
        .filter(|a| detect_agent(a.id).installed)
        .map(|a| a.id)
        .collect();

    if detected.is_empty() {
        println!("  Nothing wired.");
        println!();
        return 0;
    }

    // ── Pick agents ──
    let agent_ids: Vec<AgentId> = if !parsed.agents.is_empty() {
        let detected_set: std::collections::HashSet<AgentId> = detected.iter().copied().collect();
        parsed
            .agents
            .iter()
            .copied()
            .filter(|id| detected_set.contains(id))
            .collect()
    } else if parsed.opts.yes || !is_interactive() {
        detected.clone()
    } else {
        let select_options: Vec<crate::util::ui::SelectOption> = crate::registry::ALL_AGENTS
            .iter()
            .map(|a| {
                let is_det = detected.contains(&a.id);
                crate::util::ui::SelectOption {
                    value: a.id,
                    label: a.label.to_string(),
                    disabled: !is_det,
                    hint: if is_det {
                        "installed".to_string()
                    } else {
                        a.homepage.to_string()
                    },
                    selected: false,
                }
            })
            .collect();
        crate::util::ui::multi_select("Select agents to uninstall toksave from", select_options)
    };

    if agent_ids.is_empty() {
        println!("  Nothing selected.");
        println!();
        return 0;
    }

    // ── Pick tools ──
    let tools: Vec<ToolId> = if parsed.tools.is_empty() {
        ALL_TOOLS.iter().map(|t| t.id).collect()
    } else {
        parsed.tools.clone()
    };

    // ── Unwire ──
    let mut prog = crate::util::ui::Progress::new();
    let mut residual: Vec<(AgentId, ToolId)> = vec![];
    for agent_id in &agent_ids {
        let info = agent_info(*agent_id);
        prog.start(&format!("Uninstalling from {}", info.label));
        for tool_id in &tools {
            if !parsed.opts.dry_run {
                let _ = unwire_tool(*agent_id, *tool_id, &parsed.opts).await;
                let _ = remove_wire(
                    &format!("{:?}", agent_id).to_lowercase(),
                    tool_name(*tool_id),
                );
                // Warp's Rtk has no hook file to remove (relies on `rtk` on PATH), so
                // verify() always reports Some(true) -- that's not residual wiring.
                let is_warp_rtk = matches!((agent_id, tool_id), (AgentId::Warp, ToolId::Rtk));
                if !is_warp_rtk && verify_tool(*agent_id, *tool_id) == Some(true) {
                    residual.push((*agent_id, *tool_id));
                }
            }
        }
        prog.stop(&format!("{} {}", colors::CHECK, info.label));
    }

    for (agent_id, tool_id) in &residual {
        colors::warn(&format!(
            "{} still appears wired to {} after unwire -- config may need manual cleanup",
            tool_info(*tool_id).label,
            agent_info(*agent_id).label
        ));
    }

    // ── Cleanup cache + purge binaries on full removal ──
    if !parsed.opts.dry_run && tools.len() == ALL_TOOLS.len() && agent_ids.len() == detected.len() {
        let cache = crate::util::paths::cache_dir();
        if cache.exists() {
            let _ = std::fs::remove_dir_all(&cache);
        }
        purge_binaries_if_confirmed(parsed);
    }

    // ── Summary ──
    println!();
    let agent_labels: Vec<&str> = agent_ids.iter().map(|id| agent_info(*id).label).collect();
    let tool_labels: Vec<&str> = tools.iter().map(|id| tool_info(*id).label).collect();
    colors::ok(&format!(
        "Uninstalled {} from {}.",
        tool_labels.join(", "),
        agent_labels.join(", ")
    ));
    println!();
    0
}

fn is_interactive() -> bool {
    use std::io::IsTerminal;
    std::io::stdin().is_terminal()
}

fn tool_name(t: ToolId) -> &'static str {
    match t {
        ToolId::Rtk => "rtk",
        ToolId::Caveman => "caveman",
        ToolId::Codegraph => "codegraph",
        ToolId::ContextMode => "context-mode",
        ToolId::Ponytail => "ponytail",
        ToolId::Principles => "principles",
    }
}

/// Prompt to purge toksave-installed binaries. Mirrors TS: only asked when
/// interactive; returns false (skip) on non-TTY or user declining.
fn interactive_confirm() -> bool {
    use std::io::IsTerminal;
    if !std::io::stdin().is_terminal() {
        return false;
    }
    dialoguer::Confirm::new()
        .with_prompt("Also remove binaries/packages toksave installed (rtk, npm globals)?")
        .default(false)
        .interact()
        .unwrap_or(false)
}

fn purge_binaries_if_confirmed(parsed: &ParsedCli) {
    if parsed.opts.dry_run {
        println!("  [dry-run] would purge toksave-installed binaries + npm globals");
        return;
    }
    if std::env::var("TOKSAVE_TEST").is_ok_and(|v| v == "1") {
        return;
    }
    if !parsed.opts.yes && !interactive_confirm() {
        return;
    }

    // rtk
    let local_bin = crate::util::paths::local_bin();
    let rtk_path = if cfg!(windows) {
        local_bin.join("rtk.exe")
    } else {
        local_bin.join("rtk")
    };
    if rtk_path.exists() {
        let _ = Command::new(&rtk_path)
            .args(["init", "--uninstall"])
            .output();
        let _ = std::fs::remove_file(&rtk_path);
    }

    // npm globals
    for pkg in [
        "context-mode",
        "@colbymchenry/codegraph",
        "@dietrichgebert/ponytail",
    ] {
        let _ = Command::new(crate::util::exec::npm_cmd())
            .args(["uninstall", "-g", pkg])
            .output();
    }
}
